### PR TITLE
Integration with VSCode diagnostics + typescript

### PR DESCRIPTION
**Pull request by kujon**
Hello 👋 

This PR is my early hacked-together attempt at making this extension more user-friendly.

CleanShot 2024-04-10 at 23 34 57@2x

It does a few things:
- moves the extension code to TypeScript. Simply for speed of development.
- hooks up to VSCode's diagnostics API, so that your errors appear like any other error, rather than going into a black hole of `console.error`
- adds copied-and-pasted dead code to integrate with Code Lens in the future.

TL;DR: you'll get some nice errors in the editor, and for some types of errors, it will even highlight the correct place in the file.

My plan is to work on the following in the future (in no particular order):
- get some CI/CD to publish this to the Marketplace
- make sure that for every error, the correct place in the file is highlighted (this is probably a contribution to `vela def vet` long term, instead of hacking around regexes)
- actual tests
- extend it to work with `kind: Application` and all of the components/traits/policies deployed on the cluster you are connected to - would probably be nice to integrate with VSCode k8s extension
- Code Lens for sexier error highlighting
- any ideas from anyone to make development more productive.

### CODE REVIEW

File: validation/src/CodelensProvider.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                title: "Codelens provided by sample extension",
+                tooltip: "Tooltip provided by sample extension",
+                command: "codelens-sample.codelensAction",
+                arguments: ["Argument 1", false]
+            };
+            return codeLens;
+        }
+        return null;
+    }
+}
```
Comment (kujon) [inline validation/src/CodelensProvider.ts:54]: This whole file is dead code to be used in the future code lens integration. Copied and pasted from MS examples.

File: validation/src/extension.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    // Error: failed to parse CUE: /Users/kuba/personal_repos/vscode-plugin/validation/examples/dummy.cue: invalid type trit
+    /invalid type (.+)/,
+    // Error: failed to parse CUE: /Users/kuba/personal_repos/vscode-plugin/validation/examples/dummy.cue: invalid definition spec: json: cannot unmarshal number into Go struct field TraitDefinitionSpec.podDisruptive of type bool
+    // Error: failed to parse CUE: /Users/kuba/personal_repos/vscode-plugin/validation/examples/dummy.cue: invalid definition spec: json: cannot unmarshal number into Go struct field WorkloadTypeDescriptor.workload.type of type string
+    /field [\w\.]+\.+(\w+) of type/,
+    // Error: failed to parse CUE: /Users/kuba/personal_repos/vscode-plugin/validation/examples/dummy.cue: test2.attributes.podDisruptive: reference "tru" not found
+    /reference \"(.+)\" not found/,
+    // Error: failed to parse CUE: /Users/kuba/personal_repos/vscode-plugin/validation/examples/dummy.cue: invalid definition spec: json: unknown field "podDisruptive"
+    /unknown field "(.+)"/
+  ];
```
Comment (kujon) [inline validation/src/extension.ts:53]: ideally, all of this logic is moved over to `vela def vet` for it to report line and column numbers accurately.

**Review by FogDong — APPROVED** [review]

**Comment by FogDong** [comment]
cc @Somefive

**Review by wonderflow — APPROVED** [review]
cool

**Pull request merged**